src/fsm.cpp

```diff
@@ -171,3 +179,108 @@ void fsm::FSM::validate_final_states() const {
     }
 }
 
+bool fsm::FSM::evaluate(const char* input) {
+    fsm::String word(input);
+
+    for(int i = 0, l = word.size(); i < l; i++){
+        FSM::transition(word[i] - '0'); // convert char to int
+    }
+
+    bool flag = FSM::is_in_final_state();
+    FSM::restart();
+
+    return flag;
+}
+
+fsm::FSM fsm::FSM::operator!() const {
+    std::vector<fsm::State> newFinalStates;
+    fsm::FSM complementMachine = *this;
+
+    for(int i = 0, sz = get_states_count(); i < sz; i++){
+        if(std::find(final_states_.begin(),final_states_.end(),states_[i]) == final_states_.end()){
+            newFinalStates.push_back(states_[i]);
+        }
+    }
+
+    complementMachine.set_final_states(newFinalStates);
+
+    return complementMachine;
+}
+
+fsm::FSM fsm::FSM::operator&(const fsm::FSM &rhs) const {
+    fsm::FSM intersectionMachine;  // Initiate empty machine.
+    fsm::State comboState = get_current_state() + rhs.get_current_state();  // Get the initial state.
+    for(int i = 0, sz = get_alphabet_count(); i < sz; i++) { intersectionMachine.add_symbol(i); }  // Add the alphabet.
```

Comment:
Isn't this adding the indices of the symbols in the alphabet and not the actual symbols?

Also, shouldn't the alphabet of the intersectionMachine be the intersection of the alphabets of the two input machines?

Reply:
Yes, adding the indecies and not he actual symbols is a problem throughout that needs to be fixed.

Hmm, I am not too sure about "intersecting" machines with different alphabets, but the set analogy makes a lot of sence. I will research further.